demo/demo.js

```diff
@@ -364,3 +368,77 @@ document.getElementById("btcSignMultisig").addEventListener("click", async () =>
         }
     }
 });
+
+const cardanoGetAddress = async (display) => device.api.cardanoAddress(
+  constants.messages.CardanoNetwork.CardanoMainnet,
+  {
+    pkhSkh: {
+      keypathPayment: getKeypathFromString("m/1852'/1815'/0'/0/0"),
+      keypathStake: getKeypathFromString("m/1852'/1815'/0'/2/0"),
+    },
+  }, display);
+
+cardanoAddr.addEventListener("click", async () => {
+    try {
+        const address = await cardanoGetAddress(false);
+        cardanoAddr.textContent = "Confirm the following address on the BitBox:\n" + address;
+        cardanoAddr.className = "btn btn-warning";
+
+        await cardanoGetAddress(true);
+        cardanoAddr.textContent = "Success!:\n" + address;
+        cardanoAddr.className = "btn btn-success";
+    } catch (e) {
+        cardanoAddr.textContent = JSON.stringify(e);
+        cardanoAddr.className = "btn btn-danger";
+    }
+});
+
+cardanoPub.addEventListener("click", async () => {
+    const xpubs = await device.api.cardanoXPubs([
+        getKeypathFromString("m/1852'/1815'/0'"),
+        getKeypathFromString("m/1852'/1815'/1'"),
+    ]);
+    alert("Account #0\n" + toHex(xpubs[0]));
+    alert("Account #1\n" + toHex(xpubs[1]));
+});
+
+
+cardanoSign.addEventListener("click", async () => {
+    const ttl = "41115811";
+    const fee = "170499";
+    const response = await device.api.cardanoSignTransaction({
+        network: constants.messages.CardanoNetwork.CardanoMainnet,
+        inputs: [{
+            "keypath": getKeypathFromString("m/1852'/1815'/0'/0/0"),
+            "prevOutHash": fromHex("59864ee73ca5d91098a32b3ce9811bac1996dcbaefa6b6247dcaafb5779c2538"),
+            "prevOutIndex": 0,
+        }],
+        outputs: [{
+            "encodedAddress": "addr1q9qfllpxg2vu4lq6rnpel4pvpp5xnv3kvvgtxk6k6wp4ff89xrhu8jnu3p33vnctc9eklee5dtykzyag5penc6dcmakqsqqgpt",
+            "value": "1000000",
+        },
+         {
+             "encodedAddress": await cardanoGetAddress(false),
+             "value": "4829501",
+             "scriptConfig": {
+                 "pkhSkh": {
+                     "keypathPayment": getKeypathFromString("m/1852'/1815'/0'/0/0"),
+                     "keypathStake": getKeypathFromString("m/1852'/1815'/0'/2/0"),
+                 },
+             },
+        }],
+        fee,
+        ttl,
+        certificates: [
+            { "stakeRegistration": {
+                "keypath": getKeypathFromString("m/1852'/1815'/0'/2/0"),
+            } },
+            { "stakeDelegation": {
```

Comment:
how does this work? looks like `func (cert *cardanoCertificate) toCertificate()` will ignore delegation if registration is set

Reply:
This is an list of certificates, each of which can be one of the different certificate types. This tx has two certificates, registration and delegation.

In other words, `toCertificate()` is per certificate, and there can be many.

Reply:
yeah, i should've commented in the `toCertificate`. what if i do this:

```js
certificates: [
  {
     stakeRegistration: { ... },
     stakeDelegation: { ... },
  }
]
```

- the `toCertificate` will ignore `stakeDelegation` part
- but i will be thinking all is good
- the result is actually not what i'd probably expect

i think `toCertificate` should return an error if more than one field is specified in a single cert object.

or change types to make them explicit `stakeRegistrationCert`, `stakeDelegationCert` and so on. you can't have multiple stakeRegistration certs in a single tx anyway, can you? if so, maybe an object is a better fit for `certificates`?

```js
certificates: {
  registration: {keypath: ...},
  delegation: {keypath: ..., poolKeyhash: ...}
}
```